framework/include/cppmicroservices/BundleTracker.h

```diff
@@ -0,0 +1,474 @@
+                return trackingMap;
+            }
+            {
+                auto l = t->Lock();
+                US_UNUSED(l);
+                t->CopyEntries_unlocked(trackingMap);
+            }
+            return trackingMap;
+        }
+
+        /**
+         * Returns the tracking count for this <code>BundleTracker</code>.
+         *
+         * The tracking count is set to 0 when the <code>BundleTracker</code> is opened.
+         * The tracking count increases by 1 anytime a <code>Bundle</code> is added,
+         * modified, or removed from the <code>BundleTracker</code>.
+         * Tracking counts from different times can be compared
+         * to determine whether any bundles have changed.
+         * If the <code>BundleTracker</code> is closed, return -1.
+         *
+         * @return The current tracking count, or -1 if the <code>BundleTracker</code> is closed.
+         */
+        int
+        GetTrackingCount() const noexcept
+        {
+            auto t = d->Tracked();
+            if (!t)
+            { /* if BundleTracker has not been opened */
+                return -1;
+            }
+            {
+                auto l = t->Lock();
+                US_UNUSED(l);
+                if (t->closed)
+                { /* if BundleTracker was closed */
+                    return -1;
+                }
+                return t->GetTrackingCount();
+            }
+        }
+
+        /**
+         * Returns true if and only if this <code>BundleTracker</code> is tracking no bundles.
+         *
+         * @return true if and only if this <code>BundleTracker</code> is tracking no bundles.
+         */
+        bool
+        IsEmpty() const noexcept
+        {
+            auto t = d->Tracked();
+            if (!t)
+            { /* If BundleTracker is not open */
+                return true;
+            }
+            return (t->Lock(), t->IsEmpty_unlocked());
+        }
+
+        /**
+         * Open this <code>BundleTracker</code> to begin tracking bundles.
+         *
+         * Bundles that match the state mask will be tracked by this <code>BundleTracker</code>.
+         *
+         * @throws std::logic_error If the <code>BundleContext</code> used in the creation of this
+         * <code>BundleTracker</code> is no longer valid.
+         */
+        void
+        Open()
+        {
+            std::shared_ptr<detail::TrackedBundle<T>> t;
+            {
+                auto l = d->Lock();
+                US_UNUSED(l);
+                if (d->trackedBundle.Load() && !d->Tracked()->closed)
+                { /* If BundleTracker is open */
+                    return;
+                }
+
+                t = std::make_shared<detail::TrackedBundle<T>>(this,
+                                                               d->customizer ? d->customizer.get()
+                                                                             : d->getTrackerAsCustomizer());
```

Comment:
> This function `static_cast`'s the `this` pointer to a `BundleTrackerCustomizer<T>*` using a method in `BundleTrackerPrivate`... maybe this will be optimized out and it doesn't decrease efficiency, but is the reason we do this to keep the customizer (whether it is an actual `BundleTrackerCustomizer` or just the `BundleTracker` itself) owned by the `BundleTrackerPrivate`? It just seems at first glance a roundabout way to cast the `this` pointer.

yes, it is.

Reply:
> Additionally, if the customizer is owned by the `BundleTrackerPrivate` then I think this ternary should be abstracted to a function in the `BundleTrackerPrivate`, i.e. `GetCustomizer()` or something to keep the abstraction clear. Just feels strange for the `BundleTracker` to need to know anything about whether the `customizer` is 'valid' or just the `BundleTracker` casted

if you want to make that change, that sounds fine to me.

Reply:
Have changed it to an inline function GetCustomizer_unlocked()